### PR TITLE
Script for preprocessing the datasets for memory_saver

### PR DESCRIPTION
**Pull request by HrayrMuradyan**
## Details

A new python file is created in `audiocraft/data/` called `process_data.py`

The `DataProcessor` class of the python file is used for data preprocessing. Particularly, given the dataset paths from the checkpoint configuration, it extracts the **attributes** (namely: Tokenized Audio, Tokenized Conditions, Padding Mask) of each audio file.

The segment length of the audio is taken as 30 seconds (as provided by the checkpoint's configuration). A parameter called `time_shift` (measured in seconds) is taken to shift the 30-second window for audio clip extraction in the case of longer audios (longer than 30 seconds). Therefore, there are three possible scenarios:

- The audio is less than 30 seconds -> It is padded with 0s to become 30 seconds,
- The audio is larger or equal to 30 seconds but less than 30 seconds + `time_shift` seconds -> The first 30 seconds are taken,
- The audio is larger than 30 seconds + `time_shift` -> multiple 30 second audio clips are extracted with `time_shift` seconds shift from the full audio.

## How to run the DataProcessor
Loading the necessary libraries:
```
from audiocraft.data.process_data import PreprocessData 
from audiocraft.models.loaders import load_lm_model_ckpt
from omegaconf import OmegaConf, DictConfig
```
Reading the configuration files:
```
checkpoint = 'facebook/musicgen-small'
lm_model_ckpt = load_lm_model_ckpt(checkpoint )
cfg = OmegaConf.create(lm_model_ckpt['xp.cfg'])
```
Assigning the dataset paths:
```
cfg.datasource.train = 'egs/example'
cfg.datasource.valid = 'egs/example'
cfg.datasource.evaluate = 'egs/example'
cfg.datasource.generate = 'egs/example'
```

Instantiating the PreprocessData class:
```
preprocessor = PreprocessData(cfg, checkpoint)
```

Running the processor by providing the path for saving the tensors `save_path` and the split that should be preprocessed `data_split`:
```
preprocessor.run(time_shift=15.0, data_split='train', save_path='./dataset/tensors/')
```
The code can be found in the Jupyter Notebook file: `Data Preprocessor.ipynb`
After the successful execution of the processor, three tensors are saved for each 30-second clip from the audio:

`audio_tokens.pt` - representing the tokenized audio
`condition_tensors.pt` - representing the tokenized conditions
`padding_mask.pt` - representing the padding mask

## Notes

- Compression model used in the MusicGen solver is EncodecModel, while this version uses SEANetEncodec.

### CODE REVIEW

File: audiocraft/data/process_data.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+from audiocraft.models.encodec import EncodecModel
+
+
+class PreprocessData:
+
+	DATASET_TYPE = DatasetType.MUSIC
+
+
+	def __init__(self, cfg, checkpoint='facebook/musicgen-small'):
```
Comment (KarlosMuradyan) [inline audiocraft/data/process_data.py:39]: do you need to pass `cfg`? Seems like it is identical to `self.lm_model_ckpt`?

File: audiocraft/data/process_data.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+		conditioners = {}
+		condition_provider_args = dict_cfg.pop('args', {})
+		condition_provider_args.pop('merge_text_conditions_p', None)
+		condition_provider_args.pop('drop_desc_p', None)
+
+		for cond, cond_cfg in dict_cfg.items():
+			model_type = cond_cfg['model']
+			model_args = cond_cfg[model_type]
+			conditioners[str(cond)] = T5Conditioner(output_dim=self.cfg.transformer_lm['dim'], device=self.cfg.device, **model_args)
```
Comment (KarlosMuradyan) [inline audiocraft/data/process_data.py]: Can you confirm that `dict_cfg` will only contain entries corresponding to `T5Conditioner`? I don't think about a scenario when multiple `T5Conditioners` are initialized. 
So, my assumption would be that we only have one item in `dict_cfg` that is always `T5Conditioner`?
In that case do you want to remove the for loop?

File: audiocraft/data/process_data.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+		self.checkpoint = checkpoint
+
+		self.lm_model_ckpt = load_lm_model_ckpt(self.checkpoint)
+		self.compression_ckpt = load_compression_model_ckpt(self.checkpoint)
+
+		self.target_sr = self.cfg.sample_rate
+		self.target_channels = self.cfg.channels
+		self.segment_duration = self.cfg.dataset.segment_duration
+		self.datasets = self.build_datasets()
+		self.load_conditioner()
```
Comment (KarlosMuradyan) [inline audiocraft/data/process_data.py]: This is slightly confusing as you add an attribute 2 levels deep from this line. Can you return `condition_provider`  instead and assign it here?

File: audiocraft/data/process_data.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+		        }
+
+		output_proj_weight = state['best_state']['model'].pop('condition_provider.conditioners.description.output_proj.weight')
+		output_proj_bias = state['best_state']['model'].pop('condition_provider.conditioners.description.output_proj.bias')
+
+		conditioners['description'].output_proj.load_state_dict({'weight': output_proj_weight, 'bias': output_proj_bias})
+
+		self.condition_provider = ConditioningProvider(conditioners, device=self.cfg.device, **condition_provider_args).to(self.cfg.device)
+
+	def load_encodec(self):
```
Comment (KarlosMuradyan) [inline audiocraft/data/process_data.py:121]: Is there a way to reuse `models.builders.get_compression_model()` function here? Seems like most of the lines are from that function?

File: audiocraft/data/process_data.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+		seek_time = 0.0
+		for it, mus in enumerate(file_meta):
+			out = self.read_music(file_meta, it)
+
+			n_frames = out.shape[-1]
+			# target_frames = int(self.segment_duration * self.target_sr)
+			target_frames = int(self.segment_duration * self.target_sr)
+
+			if loader.dataset.pad:
+				padding_amount = -min(0, n_frames - target_frames)
```
Comment (KarlosMuradyan) [inline audiocraft/data/process_data.py]: Is there a reason you did `-min(0, n_frames - target_frames)` instead of `max(0, target_frames - n_frames)`?

File: audiocraft/data/process_data.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+			music_info = self.read_info(loader, meta_info, file_meta, it)
+
+			music_info.self_wav = WavCondition(
+				wav = out[None], length=torch.tensor([meta_info.n_frames]),
+				sample_rate = [meta_info.sample_rate], path=[meta_info.meta.path], seek_time=[meta_info.seek_time])
+
+			for att in loader.dataset.joint_embed_attributes:
+				att_value = getattr(music_info, att)
+				joint_embed_cond = JointEmbedCondition(
```
Comment (KarlosMuradyan) [inline audiocraft/data/process_data.py:167]: `JointEmbedCondition` is not defined

File: audiocraft/data/process_data.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+			with open(music_info_path, 'r') as json_file:
+				music_data = json.load(json_file)
+				music_data.update(info_data)
+				music_info = MusicInfo.from_dict(music_data, fields_required=loader.dataset.info_fields_required)
+			
+			if loader.dataset.paraphraser is not None:
+				music_info.description = loader.dataset.paraphraser.sample(music_info.meta.path, music_info.description)
+
+			if loader.dataset.merge_text_p:
+				music_info = augment_music_info_description(music_info, loader.dataset.merge_text_p, loader.dataset.drop_desc_p, loader.dataset.drop_other_p)
```
Comment (KarlosMuradyan) [inline audiocraft/data/process_data.py:78]: `augment_music_info_description` is not defined

File: audiocraft/data/process_data.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+			else:
+				condition_tensors, audio_tokens, padding_mask = self.prepare_attributes(out, music_info)
+				if save_path:
+					file_name = file_meta[it].path.split('\\')[-1][:-4]
+					self.save(condition_tensors, audio_tokens, padding_mask, data_split=data_split, file_name=file_name, save_path=save_path)
+					self.processing_info.save_info(file_name)
+			self.processing_info.mus_end()
+		self.processing_info.end_of_info()
+
+	def save(self, condition_tensors, audio_tokens, padding_mask, data_split, file_name, save_path):
```
Comment (KarlosMuradyan) [inline audiocraft/data/process_data.py:203]: When saving, can you call `cpu()` method? I believe it is not recommended to save tensors when their state is `cuda()`

Comment (KarlosMuradyan) [reply audiocraft/data/process_data.py:203]: and possibly `.detach()`

**Comment by KarlosMuradyan** [comment]
This is a note for later for both of us: 

Currently with this codes, only description is being encoded, without keyword.
@HrayrMuradyan I think you mentioned somewhere that keywords are being encoded with description in original codes right?

**Comment by KarlosMuradyan** [comment]
Another note is that the conditional tensors don't have the null condition

File: audiocraft/data/process_data.py (hunk 8 above)
Comment (KarlosMuradyan) [reply audiocraft/data/process_data.py:203]: and remove batch dimension. Smth like `.squeeze(0)` should do the trick, or just take `tensor[0]`, then save

File: audiocraft/data/process_data.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+					if save_path:
+						self.save(condition_tensors, audio_tokens, padding_mask, data_split=data_split, file_name=file_name_i, save_path=save_path)
+				if save_path:
+					self.processing_info.save_info(file_name, i=i)
+
+
+			else:
+				condition_tensors, audio_tokens, padding_mask = self.prepare_attributes(out, music_info)
+				if save_path:
+					file_name = file_meta[it].path.split('\\')[-1][:-4]
```
Comment (KarlosMuradyan) [inline audiocraft/data/process_data.py]: What did you want to do here? Can achieve the same thing with `os.path` module?
This didn't work for Ubuntu. Ended creating a chain of directories (actually same as the global path of the folder, smth like `save_path/home/USERNAME/....` in `save_path`.

**Comment by KarlosMuradyan** [comment]
@HrayrMuradyan functionality wise, this works well. Tested all saved files. After the above mentioned tweaks, we can merge!

File: audiocraft/data/process_data.py (hunk 1 above)
Comment (HrayrMuradyan) [reply audiocraft/data/process_data.py:39]: It is not identical. The dataset source attribute of `cfg` is changed before passing to the class (as shown in the jupyter notebook file). It was done for testing purposes, we can change it when some dataset details become clear.

File: audiocraft/data/process_data.py (hunk 2 above)
Comment (HrayrMuradyan) [reply audiocraft/data/process_data.py]: Yes, it's actually unnecessary. We can remove it in this case.

File: audiocraft/data/process_data.py (hunk 3 above)
Comment (HrayrMuradyan) [reply audiocraft/data/process_data.py]: Done!

File: audiocraft/data/process_data.py (hunk 4 above)
Comment (HrayrMuradyan) [reply audiocraft/data/process_data.py:121]: Done!

File: audiocraft/data/process_data.py (hunk 5 above)
Comment (HrayrMuradyan) [reply audiocraft/data/process_data.py]: No, those two options are identical.

File: audiocraft/data/process_data.py (hunk 6 above)
Comment (HrayrMuradyan) [reply audiocraft/data/process_data.py:167]: Done!

File: audiocraft/data/process_data.py (hunk 7 above)
Comment (HrayrMuradyan) [reply audiocraft/data/process_data.py:78]: Done!

File: audiocraft/data/process_data.py (hunk 8 above)
Comment (HrayrMuradyan) [reply audiocraft/data/process_data.py:203]: Done!

File: audiocraft/data/process_data.py (hunk 9 above)
Comment (HrayrMuradyan) [reply audiocraft/data/process_data.py]: Done! Please confirm later if it worked for your OS.

Comment (KarlosMuradyan) [reply audiocraft/data/process_data.py]: Yes it worked, thanks

**Comment by KarlosMuradyan** [comment]
Looks good. Merging!
